flow/aggregates.py

```diff
@@ -0,0 +1,266 @@
+            raise TypeError(f"Expected callable for select, got {type(select)}")
+
+        if sort_by is None and select is None and not reverse_order:
+            self._is_aggregate = getattr(aggregator, '_num', 0) != 1
+        else:
+            self._is_aggregate = True
+
+        self._aggregator = aggregator
+        self._sort_by = sort_by
+        self._reverse_order = reverse_order
+        self._select = select
+
+    @classmethod
+    def groupsof(cls, num=1, sort_by=None, reverse_order=False, select=None):
+        # copied from: https://docs.python.org/3/library/itertools.html#itertools.zip_longest
+        try:
+            num = int(num)
+            if num <= 0:
+                raise ValueError('The num parameter should have a value greater than 0')
+        except TypeError:
+            raise TypeError('The num parameter should be an integer')
+
+        def aggregator(jobs):
+            args = [iter(jobs)] * num
+            return zip_longest(*args)
+
+        setattr(aggregator, '_num', num)
+
+        return cls(aggregator, sort_by, reverse_order, select)
+
+    @classmethod
+    def groupby(cls, key, default=None, sort_by=None, reverse_order=False, select=None):
+        if isinstance(key, str):
+            if default is None:
+                def keyfunction(job):
+                    return job.sp[key]
+            else:
+                def keyfunction(job):
+                    return job.sp.get(key, default)
+
+        elif isinstance(key, Iterable):
+            keys = list(key)
+
+            if default is None:
+                def keyfunction(job):
+                    return [job.sp[key] for key in keys]
+            else:
+                if isinstance(default, Iterable):
+                    if len(default) != len(keys):
+                        raise ValueError("Expected length of default argument is "
+                                         f"{len(keys)}, got {len(default)}.")
+                else:
+                    raise TypeError("Invalid default argument. Expected Iterable, "
+                                    f"got {type(default)}")
+
+                def keyfunction(job):
+                    return [job.sp.get(key, default_value)
+                            for key, default_value in zip(keys, default)]
+
+        elif callable(key):
+            keyfunction = key
+
+        else:
+            raise TypeError("Invalid key argument. Expected either str, Iterable "
+                            f"or a callable, got {type(key)}")
+
+        def aggregator(jobs):
+            for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
+                yield group
+
+        return cls(aggregator, sort_by, reverse_order, select)
+
+    def __hash__(self):
+        blob_l = map(hash, [self._aggregator, self._sort_by,
+                            self._reverse_order, self._select])
+        blob = ','.join(str(attr) for attr in blob_l)
```

Comment:
To use an object as a key in a dictionary it has to support `__hash__` and `__eq__`

Reply:
Just FYI @kidrahahjo both of these methods have default implementations for any class (e.g. `class A: pass` will have default versions of both methods defined), but I assume @b-butler wants you to define a meaningful equality check here.

Reply:
That is correct.